File: aoc2023/day1b.py
import sys


def kmp_search(text: str, pattern: str, table: list[int] | None = None) -> list[int]:
    """
    Gets the start indices of all matches of pattern inside text
    @param text String to search for occurrences of pattern
    @param pattern Substring to search occurrences of within text
    @param Partial match table computed by kmp_preprocess
    @return Array of match indices
    """
    if table is None:
        table = kmp_preprocess(pattern)

    result: list[int] = []
    j = 0
    k = 0

    while j < len(text):
        if pattern[k] == text[j]:
            j += 1
            k += 1
            if k == len(pattern):
                result.append(j - k)
                k = table[k]
        else:
            k = table[k]
            if k < 0:
                j += 1
                k += 1

    return result


def kmp_preprocess(pattern: str) -> list[int]:
    """
    Computes partial match table for KMP algorithm
    @param pattern Pattern for which to compute partial match table
    @return Table of indices to continue search at
    """
    result = [-1] * (len(pattern) + 1)
    candidate = 0

    for i in range(1, len(pattern)):
        if pattern[i] == pattern[candidate]:
            result[i] = result[candidate]
        else:
            result[i] = candidate
            while candidate >= 0 and pattern[i] != pattern[candidate]:
                candidate = result[candidate]
        candidate += 1

    result[-1] = candidate

    return result


def run():
    lines: list[int] = []
    patterns = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
    tables = [kmp_preprocess(x) for x in patterns]
    values = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    for line in sys.stdin:
        digits: list[tuple[int, int]] = []  # Tuples of index, value
        digits.extend([(i, int(c)) for i, c in enumerate(line) if c.isnumeric()])

        for table, value, pattern in zip(tables, values, patterns):
            digits.extend([(i, value) for i in kmp_search(line, pattern, table)])
        digits.sort()

        lines.append(digits[0][1] * 10 + digits[-1][1])

    print(sum(lines))


if __name__ == "__main__":
    run()

File: src/aoc2023/day1b.rs
use std::io::BufRead;

pub fn kmp_preprocess(pattern: &str) -> Vec<i32> {
    let mut result = vec![-1; pattern.len() + 1];
    let mut candidate: i32 = 0;

    for i in 1..pattern.len() {
        if pattern[i..i + 1] == pattern[candidate as usize..candidate as usize + 1] {
            result[i] = result[candidate as usize]
        } else {
            result[i] = candidate;

            while candidate >= 0
                && pattern[i..i + 1] != pattern[candidate as usize..candidate as usize + 1]
            {
                candidate = result[candidate as usize];
            }
        }

        candidate += 1;
    }
    result[pattern.len()] = candidate;

    result
}

pub fn kmp_search(text: &str, pattern: &str, table: Option<&Vec<i32>>) -> Vec<usize> {
    let mut result: Vec<usize> = Vec::new();
    let mut j: i32 = 0;
    let mut k: i32 = 0;

    let lookup: &Vec<i32>;
    let lookup_table: Vec<i32>;
    match table {
        Some(x) => {
            lookup = x;
        }
        _ => {
            lookup_table = kmp_preprocess(pattern);
            lookup = &lookup_table;
        }
    }

    while j < text.len() as i32 {
        if text[j as usize..j as usize + 1] == pattern[k as usize..k as usize + 1] {
            j += 1;
            k += 1;
            if k == pattern.len() as i32 {
                result.push((j - k) as usize);
                k = lookup[k as usize];
            }
        } else {
            k = lookup[k as usize];
            if k < 0 {
                j += 1;
                k = 0;
            }
        }
    }

    result
}

pub fn run() {
    let patterns = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];
    let values: Vec<u32> = (1..=9).collect();
    let tables: Vec<Vec<i32>> = patterns.iter().map(|x| kmp_preprocess(x)).collect();

    let result: Vec<u32> = std::io::stdin()
        .lock()
        .lines()
        .map(|x| {
            let text = x.unwrap();

            let mut y: Vec<(usize, u32)> = patterns
                .iter()
                .zip(values.iter())
                .zip(tables.iter())
                .flat_map(|((pattern, value), table)| {
                    kmp_search(&text, pattern, Some(table))
                        .into_iter()
                        .map(|index| (index, *value))
                })
                .collect();

            let digits: Vec<(usize, u32)> = text
                .chars()
                .enumerate()
                .filter(|(_, y)| y.is_numeric())
                .map(|(i, y)| (i, y.to_digit(10).unwrap()))
                .collect();

            y.extend(digits);

            y.sort();

            y[0].1 * 10 + y[y.len() - 1].1
        })
        .collect();

    let final_sum: u32 = result.into_iter().sum();
    println!("{final_sum}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kmp_preprocess() {
        assert_eq!(
            kmp_preprocess("abacababc"),
            vec![-1, 0, -1, 1, -1, 0, -1, 3, 2, 0]
        );

        assert_eq!(kmp_preprocess("abcdabd"), vec![-1, 0, 0, 0, -1, 0, 2, 0]);
    }

    #[test]
    fn test_kmp_search() {
        assert_eq!(kmp_search("asdfasdf", "asdf", None), vec![0, 4]);

        assert_eq!(kmp_search("zerone", "one", None), vec![3]);
    }
}
